### ISSUE TITLE
Zero-shot Code

### ISSUE THREAD

**Issue by zongzi3zz** [kuai-lab/sound-guided-semantic-image-manipulation#11]
Thanks for the great work! Can you share the code of zero-shot transfer on the ESC-50 and  Urban sound 8k?  Thanks.

**Comment by wqliu657** [comment]
111

**Comment by lsh3163** [comment]
Thanks for your interest!! 

This is our raw version of measuring the accuracy of the zero-shot transfer. 

```python
for row in df.iterrows():
    idx, row = row
    fold = row["fold"]
    if fold == test_fold:
        test_size += 1
    if dataset_name=="esc":
        filename = row["filename"]
        category = row["category"]
        filepath = "./ESC-50-master/audio/"+filename
    elif dataset_name=="urban":
        filename = row["slice_file_name"]
        category = row["class"]
        filepath = f"./urban8k/fold{fold}/"+filename
        
    y, sr = librosa.load(filepath, sr=44100)
    audio_inputs = librosa.feature.melspectrogram(y=y, sr=sr, n_mels=n_mels)
    audio_inputs = librosa.power_to_db(audio_inputs, ref=np.max) / 80.0 + 1
    zero = np.zeros((n_mels, time_length))

    h, w = audio_inputs.shape
    if w >= time_length:
        j = (w - time_length) // 2
        audio_inputs = audio_inputs[:,j:j+time_length]
    else:
        j = (time_length - w) // 2
        zero[:,j:j+w] = audio_inputs[:,:w]
        audio_inputs = zero
    
    audio_inputs = cv2.resize(audio_inputs, (n_mels, resize_resolution))
    audio_inputs = np.array([audio_inputs])
    audio_inputs = torch.from_numpy(audio_inputs.reshape((1, 1, n_mels, resize_resolution))).float().to(device)

    with torch.no_grad():
        text_tokens = torch.cat([clip.tokenize(text) for text in labels])
        text_embedding = clip_model.encode_text(text_tokens.to(device)).float()

        class_tokens = torch.cat([clip.tokenize(text) for text in [category]])
        class_embedding = clip_model.encode_text(class_tokens.to(device)).float()

        audio_embedding = audio_encoder(audio_inputs)
        
        audio_embedding = audio_embedding / audio_embedding.norm(dim=-1, keepdim=True)
        text_embedding = text_embedding / text_embedding.norm(dim=-1, keepdim=True)

        class_embedding = class_embedding / class_embedding.norm(dim=-1, keepdim=True)
        
        
        if fold==test_fold:
            x_train.append(audio_embedding.cpu().numpy().tolist()[0])
            y_train.append(category)
        elif fold!=test_fold:
            x_test.append(audio_embedding.cpu().numpy().tolist()[0])
            y_test.append(category)

        proj_per_audio = (audio_embedding @ text_embedding.T) * math.exp(0.07)
        label_idx = torch.argmax(proj_per_audio, axis=1)
        pred_category = labels[label_idx]
        if True:
            modal_tsne.append(class_embedding.cpu().numpy()[0])
            domain_labels.append(category)
            modal_tsne.append(audio_embedding.cpu().numpy()[0]* math.exp(0.07))
    print(idx, pred_category, category, len(text_embedding.cpu().numpy().tolist()[0]), proj_per_audio[0][label_idx])
    
    all_predictions.append(proj_per_audio[0][label_idx].cpu().numpy().tolist()[0])
    if pred_category == category:
        all_gts.append(1)
    else:
        all_gts.append(0)
    all_size += 1
    # if pred_category == category and fold==test_fold:
    #     correct += 1

    if pred_category == category:
        correct += 1
```

**Comment by zongzi3zz** [comment]
Thanks @lsh3163 for the quick response！
